### PR TITLE
Add RichText support to find.text()

### LINKED ISSUE #21961: Support RichText in finders
Internal: b/193847760

Let's make this test pass:

```dart
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(body: RichText(text: TextSpan(text: "MyRichText"))),
    ));
    expect(find.text("MyRichText"), findsOneWidget);
```

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
## Description

Adds support for matching standalone `RichText` widgets to `find.text()` in a **backwards-compatible** manner. That is, this does not break any existing tests (which https://github.com/flutter/flutter/pull/21964 did).

It adds the arguably better strategy for matching `Text` widgets (using only `RichText`) behind a `skipRichText` flag that has to be set to `false`.  

### Documentation

I have added both extensive additions to the existing `find.text()` docs as well as a comment to explain why we are using the inferior `widget is Text` method of looking up `Text` widgets (when we could just always **only** look for `RichText` as that is also inserted by `Text` widgets).

### Tests

I have added test cases for every possible scenario, which ensures that both the new `skipRichText: false` strategy finds every text widget **once**, as well as testing that the old strategy does not match `RichText` at all.  
The semantics test case is taken from https://github.com/flutter/flutter/pull/21964.

## Issues

This resolves a popular issue:

* Closes #21961.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by creativecreatorormaybenot** [comment]
cc @mehmetf @jonahwilliams

File: packages/flutter_test/lib/src/finders.dart [hunk 1]
```diff
@@ -548,26 +572,66 @@ abstract class MatchFinder extends Finder {
...
+      return widget.text.toPlainText() == text;
+    return false;
+  }
+
+  bool _matchesNonRichText(Widget widget) {
     if (widget is Text) {
       if (widget.data != null)
         return widget.data == text;
       assert(widget.textSpan != null);
       return widget.textSpan!.toPlainText() == text;
```
Comment (creativecreatorormaybenot) [inline packages/flutter_test/lib/src/finders.dart:627]: To further explain: in my opinion, this code is completely **redundant** as directly matching `RichText` is the way better approach (as all `Text` widgets insert a `RichText` child).

However, we sadly **have** to keep this old code as we would otherwise break user code...

File: packages/flutter_test/lib/src/finders.dart [hunk 2]
```diff
@@ -23,18 +23,42 @@ const CommonFinders find = CommonFinders._();
 class CommonFinders {
   const CommonFinders._();
 
-  /// Finds [Text] and [EditableText] widgets containing string equal to the
-  /// `text` argument.
+  /// Finds [Text], [EditableText], and optionally [RichText] widgets
+  /// containing string equal to the [text] argument.
+  ///
+  /// If [skipRichText] is disabled, [RichText] widgets (and therefore also
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:29]: here and elsewhere: method prams like skipRichText should be surrounded with `` instead of [].

File: packages/flutter_test/lib/src/finders.dart [hunk 3]
```diff
@@ -23,18 +23,42 @@ const CommonFinders find = CommonFinders._();
...
+  /// ```
+  ///
+  /// This will match [Text], [Text.rich], [EditableText], as well as standalone
+  /// [RichText] widgets that contain the "Close" string.
+  Finder text(
+    String text, {
+    bool skipRichText = true,
+    bool skipOffstage = true,
+  }) =>
+      _TextFinder(text, skipRichText: skipRichText, skipOffstage: skipOffstage);
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:61]: nit: keep this on one line or change it to a multi-line `{ return ... }` block

File: packages/flutter_test/lib/src/finders.dart [hunk 4]
```diff
@@ -548,26 +572,66 @@ abstract class MatchFinder extends Finder {
...
 
   final String text;
 
+  /// Whether standalone [RichText] widgets should be skipped or not.
+  ///
+  /// Defaults to `true`.
+  ///
+  /// If enabled, only [Text] widgets will be considered. These of course also
+  /// insert a [RichText] widget, however, that is ignored and so are all
+  /// standalone [RichText] widgets.
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:589]: This sentence is hard to understand. Maybe rephrase?

File: packages/flutter_test/lib/src/finders.dart [hunk 5]
```diff
@@ -548,26 +572,66 @@ abstract class MatchFinder extends Finder {
...
+  /// If disabled, only [RichText] widgets will be considered. This *implicitly*
+  /// considers [Text] widgets as well since they always insert a [RichText]
+  /// child.
+  ///
+  /// In either case, [EditableText] widgets will also be considered.
+  final bool skipRichText;
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:595]: maybe we should name this positively: findRichText or something like that.

Comment (creativecreatorormaybenot) [reply packages/flutter_test/lib/src/finders.dart:595]: @goderbauer I thought about this as well, but I wanted to match `skipOffstage`. What do you think?

File: packages/flutter_test/lib/src/finders.dart (hunk 2 above)
Comment (creativecreatorormaybenot) [reply packages/flutter_test/lib/src/finders.dart:29]: Thanks! I am wondering, is there a particular reason for this in dartdoc? In my IDE, I can use code navigation on parameters (so `skipRichText` correctly links to the parameter).

File: packages/flutter_test/lib/src/finders.dart (hunk 4 above)
Comment (creativecreatorormaybenot) [reply packages/flutter_test/lib/src/finders.dart:589]: Sure, I can see it :)

File: packages/flutter_test/lib/src/finders.dart (hunk 3 above)
Comment (creativecreatorormaybenot) [reply packages/flutter_test/lib/src/finders.dart:61]: Sounds good, I thought about how to do it but I did not see any blocks elsewhere, but it for sure is the better style choice.

**Comment by wanglikun7342** [comment]
@jonahwilliams  please review this pr

**Comment by creativecreatorormaybenot** [comment]
Hi @goderbauer - I addressed all of your concerns. Can you review again? :)

File: packages/flutter_test/lib/src/finders.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter_test/lib/src/finders.dart:599]: I think skipOffstage is also poorly named. Since we are adding something new, let's give it a better name.

File: packages/flutter_test/lib/src/finders.dart [hunk 6]
```diff
@@ -23,18 +23,47 @@ const CommonFinders find = CommonFinders._();
...
-  /// Finds [Text] and [EditableText] widgets containing string equal to the
-  /// `text` argument.
+  /// Finds [Text], [EditableText], and optionally [RichText] widgets
+  /// containing string equal to the `text` argument.
+  ///
+  /// If `skipRichText` is disabled, [RichText] widgets (and therefore also
+  /// [Text] and [Text.rich] widgets) are matched by comparing the
+  /// [InlineSpan.toPlainText] with the given `text`.
+  /// If `skipRichText` is enabled, all standalone [RichText] widgets are
+  /// ignored and [Text] matching shortcuts to [Text.data].
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:33]: This sounds confusing, maybe better:

```suggestion
  /// ignored and `text` is matched with [Text.data].
```

**Comment by creativecreatorormaybenot** [comment]
@goderbauer I addressed your comments from code review.

**Comment by goderbauer** [comment]
Looks the analyzer is unhappy and some tests are failing. Can you take a look?

**Comment by creativecreatorormaybenot** [comment]
@goderbauer Thanks - I cannot believe I missed that; updated it.

File: packages/flutter_test/lib/src/finders.dart [hunk 7]
```diff
@@ -23,18 +23,47 @@ const CommonFinders find = CommonFinders._();
...
   const CommonFinders._();
 
-  /// Finds [Text] and [EditableText] widgets containing string equal to the
-  /// `text` argument.
+  /// Finds [Text], [EditableText], and optionally [RichText] widgets
+  /// containing string equal to the `text` argument.
+  ///
+  /// If `findRichText` is disabled, all standalone [RichText] widgets are
+  /// ignored and `text` is matched with [Text.data] or [Text.textSpan].
+  /// If `findRichText` is enabled, [RichText] widgets (and therefore also
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/finders.dart:31]:
```suggestion
  /// If `findRichText` is true, [RichText] widgets (and therefore also
```

**Review by goderbauer — APPROVED** [review]
LGTM after nits are fixed.

**Comment by creativecreatorormaybenot** [comment]
@goderbauer done 👍

**Comment by creativecreatorormaybenot** [comment]
@dnfield thx a lot ❤️

File: packages/flutter_test/lib/src/finders.dart (hunk 2 above)
Comment (bartekpacia) [reply packages/flutter_test/lib/src/finders.dart]: I'd also love to know the reason behind this.
